### PR TITLE
Improve handler api

### LINKED ISSUE #103: SMCHandler/ShardTracker API improvements
Here are some ideas for improvements of the API of `SMCHandler` and `ShardTracker` I got when implementing `SMCService` (most of them for the `__init__` method and some probably a little bit subjective):

1) I think `SMCHandler` should be renamed to just `SMC` as its a subclass of `web3.Contract`. While it technically does not refer to the contract itself, `SMC` would be more inline with web3's terminology (contract <- sharding manager contract).

2) Both handler and tracker take a `config` dictionary which mainly contains parameters for the contract. I think it would be better if those parameters would be retrieved from the on-chain contract itself, otherwise it's possible that there is a mismatch. The remaining two ("DEFAULT_GAS" and "GAS_PRICE") could just be passed as standalone parameters or removed (see next bullet point).

3) The various `send_transaction` methods in `SMCHandler` take a `gas` argument. As the required gas of these functions is known (it's part of the ABI) it would be nice if it would just use that number.

4) `ShardTracker` expect a `LogHandler` which could just be created by the `ShardTracker` itself.

5) To create the `SMCHandler` one needs to provide the ABI. As the ABI doesn't change that parameter is useless, the constructor could just load the ABI from JSON.

6) Deploying the SMC is quite cumbersome. Two ideas here: Either make the test utils importable for other packages or, even better, make deploying easy by setting the bytecode. Then one can just use the `constructor` functionality provided by `web3` (if I understand the docs correctly, haven't actually tried it).

@NIC619 what do you think of these? I would have already started making PRs, but wanted to discuss first, and also give you the chance to work on them if you like. I'm mainly blocked by the last point because there's not really a good way to test the `SMCService` right now.

### PR DESCRIPTION
**Pull request by NIC619**
### What was wrong?
Fix #103 , #107, #111 

### How was it fixed?
- [x] #103, handler api improvement
    - [x] Rename `SMCHandler` to `SMC`
    - [x] Instantiate `LogHandler` in `ShardTracker` instead of passing in an instance
    - [x] Remove `gas` option and `DEFAULT_GAS` in config*
    - [ ] ~Look into new signing middleware in web3~
- [x] #107, fix mismatch config name between SMC and config
- [x] #111, convert exposed address to canonical form

Note:
*The estimate gas utils provided by py-evm doesn't seems to work steadily. Some transaction fails while some doesn't. Haven't have time to dig into it so currently I use estimated gas extracted from ABI  for transaction and estimated gas utils from py-evm for local call.

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by NIC619** [comment]
@hwwhww @jannikluhn 
I've taken a look into the new signing middleware in web3 and `SMCService` seems like a better place for the setup so I decided to remove it from this PR and it's ready for review now.

File: sharding/handler/smc_handler.py [hunk 1]
```diff
@@ -34,25 +35,33 @@
...
     abi = smc_json["abi"]
     bytecode = decode_hex(smc_json["bytecode"])
 
-    private_key = None  # type: datatypes.PrivateKey
-    sender_address = None  # type: Address
+    default_priv_key = None  # type: datatypes.PrivateKey
+    default_sender_address = None  # type: Address
     config = None  # type: Dict[str, Any]
 
+    _estimate_gas_dict = dict()  # type: Dict[str, Any]
```
Comment (jannikluhn) [inline sharding/handler/smc_handler.py]: As it's the same for all instances, we could populate it here already instead of in `__init__` (or even outside the class and only reference it here).

Comment (jannikluhn) [reply sharding/handler/smc_handler.py]: And the type should be `Dict[str, int]` instead of `Dict[str, Any]`.

File: sharding/handler/utils/log_parser.py [hunk 2]
```diff
@@ -66,7 +66,7 @@ def _parse_value(self, *, val_type: str, val: bytes) -> Union[bool, str, bytes,
         if val_type == 'bool':
             return bool(big_endian_to_int(val))
         elif val_type == 'address':
-            return to_checksum_address(val[-20:])
+            return to_canonical_address(val[-20:])
```
Comment (jannikluhn) [inline sharding/handler/utils/log_parser.py]: As this returns an `Address`, the type hint of the function should be updated (replace `str` with `Address`). Unfortunately, `eth_utils` does not use `eth_typing` yet, so they defined their own `Address` type. Therefore it might be necessary to `cast` their `Address into `eth_typing.Address`, I'm not sure though.

File: tests/contract/test_log_emission.py [hunk 3]
```diff
@@ -36,7 +32,7 @@ def test_log_emission(smc_handler):  # noqa: F811
     # Check that log was successfully emitted
     log = shard_tracker.get_register_notary_logs()[0]
     assert getattr(log, 'index_in_notary_pool') == 0 and \
```
Comment (jannikluhn) [inline tests/contract/test_log_emission.py]: Not part of the PR, but why do we use `getattr` instead of just `log.index_in_notary_pool` here?

**Comment by jannikluhn** [comment]
> I've taken a look into the new signing middleware in web3 and SMCService seems like a better place for the setup so I decided to remove it from this PR and it's ready for review now.

Assuming that we use the signing middleware in `SMCService` do we still need `SMC.default_priv_key`? There doesn't seem to be documentation on that part of web3.py yet.

File: sharding/handler/smc_handler.py [hunk 4]
```diff
@@ -34,25 +35,33 @@
...
-        self.sender_address = self.private_key.public_key.to_canonical_address()
+        self.default_priv_key = default_priv_key
+        self.default_sender_address = self.default_priv_key.public_key.to_canonical_address()
         self.config = config
 
+        # Extract estimate gas from abi
+        for entry in self.abi:
+            if entry['type'] == 'function':
+                if not entry['constant']:
+                    self._estimate_gas_dict[entry['name']] = entry['gas']
```
Comment (jannikluhn) [inline sharding/handler/smc_handler.py]: This can be simplified:
```
self._estimate_gas_dict = {
    entry['name']: entry['gas']
    for entry in self.abi
    if entry['type'] == 'function' and not entry['constant']
}
```
